src/channel_manager.ts

```diff
@@ -227,24 +256,85 @@ export class ChannelManager extends WithSubscriptions {
     this.options = { ...DEFAULT_CHANNEL_MANAGER_OPTIONS, ...options };
   };
 
+  private queryChannelsRequest = async (
+    payload: QueryChannelsRequestPayload,
+    retryCount = 0,
+  ): Promise<void> => {
+    const { filters, sort, options, stateOptions } = payload;
+    const { offset, limit } = {
+      ...DEFAULT_CHANNEL_MANAGER_PAGINATION_OPTIONS,
+      ...options,
+    };
+    try {
+      const channels = await this.client.queryChannels(
+        filters,
+        sort,
+        options,
+        stateOptions,
+      );
+      const newOffset = offset + (channels?.length ?? 0);
+      const newOptions = { ...options, offset: newOffset };
+      const { pagination } = this.state.getLatestValue();
+
+      this.state.partialNext({
+        channels,
+        pagination: {
+          ...pagination,
+          hasNext: (channels?.length ?? 0) >= limit,
+          isLoading: false,
+          options: newOptions,
+        },
+        initialized: true,
+        error: undefined,
+      });
+      this.client.offlineDb?.executeQuerySafely(
+        (db) =>
+          db.upsertCidsForQuery({
+            cids: channels.map((channel) => channel.cid),
+            filters: pagination.filters,
+            sort: pagination.sort,
+          }),
+        { method: 'upsertCidsForQuery' },
+      );
+    } catch (err) {
+      if (retryCount >= DEFAULT_QUERY_CHANNELS_RETRY_COUNT) {
+        console.warn(err);
+
+        const wrappedError = new Error(
+          `Maximum number of retries reached in queryChannels. Last error message is: ${err}`,
+        );
+
+        this.state.partialNext({ error: wrappedError });
+        return;
+      }
+
+      await waitSeconds(DEFAULT_QUERY_CHANNELS_SECONDS_BETWEEN_RETRIES);
+
+      return this.queryChannelsRequest(payload, retryCount + 1);
+    }
+  };
+
   public queryChannels = async (
     filters: ChannelFilters,
     sort: ChannelSort = [],
     options: ChannelOptions = {},
     stateOptions: ChannelStateOptions = {},
   ) => {
-    const { offset, limit } = {
-      ...DEFAULT_CHANNEL_MANAGER_PAGINATION_OPTIONS,
-      ...options,
-    };
     const {
-      pagination: { isLoading },
+      pagination: { isLoading, filters: filtersFromState },
+      initialized,
     } = this.state.getLatestValue();
 
-    if (isLoading && !this.options.abortInFlightQuery) {
+    if (
+      isLoading &&
+      !this.options.abortInFlightQuery &&
+      JSON.stringify(filtersFromState) === JSON.stringify(filters)
```

Comment:
Does it make sense to rely on order of keys in objects?

Reply:
Perhaps not, ideally some way of hashing would be nice here I guess; however filters are mostly fire and forget on the side of the SDKs anyway (and the SDKs themselves don't really handle property ordering either) so we can add this to the bucket of future improvements.

I'll add a `TODO` though so we don't forget.